### PR TITLE
Replace Deprecated Mail::Sender by Email::Sender

### PR DESCRIPTION
**Pull request by tkempf**
Replaces the deprecated Mail::Sender with Email::Sender without the need of changing master.cf.
Works with TLS on Debian Stretch
Adds a regexp novacation_pattern to define adresses where vacation E-mails never should be sent

### CODE REVIEW

**Comment by DavidGoodwin** [comment]
Thank you. 

Can you include a comment near where $novacation_pattern is specified which illustrates what it could look like?

File: VIRTUAL_VACATION/vacation.pl [hunk 1]
```diff
@@ -86,25 +81,32 @@
 #             Add capability to include the subject of the original mail in the subject of the vacation message.
 #             A good vacation subject could be: 'Re: $SUBJECT'
 #             Also corrected log entry about "Already informed ..." to show the $orig_from, not $email
+# 
+# 2017-07-14  Thomas Kempf <tkempf@hueper.de>
+#			  Replacing deprecated Mail::Sender by Email::Sender
+#			  Add configuration parameter $novacation_pattern in order to exlude specific alias-recipients from 
+# 			  sending vacation mails, even if one or multiple of the recipients the alias points to has vacation
+#             currently active.
```
Comment (cboltz) [inline VIRTUAL_VACATION/vacation.pl]: You have a mix of tabs and spaces here. Please use only spaces.

File: VIRTUAL_VACATION/vacation.pl [hunk 2]
```diff
@@ -86,25 +81,32 @@
...
 # You may install these via CPAN, or through your package tool.
 # CPAN: 'perl -MCPAN -e shell', then 'install Module::Whatever'
 #
 # On Debian based systems :
-#   libmail-sender-perl
-#   libdbd-pg-perl
+#   libemail-sender-perl
+#   libemail-simple-perl
 #   libemail-valid-perl
+#	libtry-tiny-perl
```
Comment (cboltz) [inline VIRTUAL_VACATION/vacation.pl]: another tab that should be a space ;-)

File: VIRTUAL_VACATION/vacation.pl [hunk 3]
```diff
@@ -207,6 +215,15 @@
...
 our $noreply_pattern = 'bounce|do-not-reply|facebook|linkedin|list-|myspace|twitter'; 
 
+# Never send vacation mails for the following recipient email addresses.
+# Useful for e.g. aliases pointing to multiple recipients which have vacation active
+# hence an email to the alias should not trigger vacation messages.
+# By default vacation email addresses will be sent for all recipients.
+# default = ''
+# preventing vacation notifications for recipient info@example.org would look like this:
+# our $novacation_pattern = 'info\@example\.org';
+our $novacation_pattern = '';
```
Comment (cboltz) [inline VIRTUAL_VACATION/vacation.pl]: Good idea!

Nitpicking: could you please use a slightly more readable variable name, maybe $no_vacation_pattern?

File: VIRTUAL_VACATION/vacation.pl [hunk 4]
```diff
@@ -536,47 +556,53 @@ sub send_vacation_email {
...
+		$email = Email::Simple->create(
+            header => [
+                To      => $to,
+                From    => $from,
+                Subject => $subject,
+                Precedence => 'junk',
+                'Content-Type' => "text/plain; charset=utf-8",
+                'X-Loop' => 'Postfix Admin Virtual Vacation',
+            ],
+            body => $body,
```
Comment (cboltz) [inline VIRTUAL_VACATION/vacation.pl]: This section contains a wild mix of tabs and spaces. Please change it to spaces only.

Besides that, the new code looks good (disclaimer: I only read it, but didn't test it)

File: VIRTUAL_VACATION/vacation.pl [hunk 5]
```diff
@@ -536,47 +556,53 @@ sub send_vacation_email {
...
+        try {
+            sendmail($email, { transport => $transport });
+        } finally {
+            if (@_) {
+                $logger->error("Failed to send vacation response to $to from $from subject $subject: @_");
+            } else {
+            	$logger->debug("Vacation response sent to $to from $from subject $subject sent\n");
+        	}
+    	}
+	}
```
Comment (cboltz) [inline VIRTUAL_VACATION/vacation.pl]: more tabs ;-)

File: VIRTUAL_VACATION/vacation.pl [hunk 6]
```diff
@@ -695,6 +722,12 @@ sub check_and_clean_from_address {
     exit(0);
 }
 $logger->debug("Email headers have to: '$to' and From: '$from'");
+
+if ($to =~ /^.*($novacation_pattern).*/i) {
```
Comment (Callidior) [inline VIRTUAL_VACATION/vacation.pl]: With the empty default value for `$novacation_pattern`, this will match everything, prevent vacation mails from ever being sent.

File: VIRTUAL_VACATION/vacation.pl [hunk 7]
```diff
@@ -536,47 +556,53 @@ sub send_vacation_email {
...
+            $logger->info("Doing SASL Authentication with user $smtp_params->{sasl_username}\n");
+		};
+		
+        my $transport = Email::Sender::Transport::SMTP->new($smtp_params);
+		
+		$email = Email::Simple->create(
+            header => [
+                To      => $to,
+                From    => $from,
+                Subject => $subject,
```
Comment (Callidior) [inline VIRTUAL_VACATION/vacation.pl]: Why did you remove `encode_mimewords($subject, 'Charset', 'UTF-8')`? This way, UTF-8 characters in vacation subjects are not encoded as quoted printable and will be rendered incorrectly.

Comment (tkempf) [inline VIRTUAL_VACATION/vacation.pl]: Because i didn't read the RFCs and tests with non ASCII-chars in the Subject worked with my client.   I'll revert that change.

**Comment by DavidGoodwin** [comment]
I tested this locally; test.sh didn't explode. I'll move to using it on a real server and see how that fares. Thank you!
